kythe/typescript/indexer.ts

```diff
@@ -543,83 +501,153 @@ class Visitor {
+      root: vname && vname.corpus ? vname.root : this.compilationUnit.root,
+      path: vname && vname.path ? vname.path : path,
+    };
+  }
 
-    if (!contextApplies && vnames && vnames[ns]) return vnames[ns]!;
-    // TODO: update symbolNames table to account for context kind
+  /**
+   * emit emits a Kythe entry, structured as a JSON object.  Defaults to
+   * emitting to stdout but users may replace it.
+   */
+  emit = (obj: JSONFact|JSONEdge) => {
+    console.log(JSON.stringify(obj));
+  };
 
-    if (!declarations || declarations.length < 1) {
-      throw new Error('TODO: symbol has no declarations?');
+  /**
+   * anonName assigns a freshly generated name to a Node.
+   * It's used to give stable names to e.g. anonymous objects.
+   */
+  private anonName(node: ts.Node): string {
+    let name = this.anonNames.get(node);
+    if (!name) {
+      name = `anon${this.anonId++}`;
+      this.anonNames.set(node, name);
     }
+    return name;
+  }
+}
 
-    // Disambiguate symbols with multiple declarations using a context. This
-    // only applies to getters and setters currently.
-    if (contextApplies) {
-      switch (context) {
-        case Context.Getter:
-          declarations = declarations.filter(ts.isGetAccessor);
-          break;
-        case Context.Setter:
-          declarations = declarations.filter(ts.isSetAccessor);
-          break;
-      }
-    }
-    // Otherwise, if there are multiple declarations but no context is
-    // provided, try to return the getter declaration.
-    else if (declarations.length > 1) {
-      const getDecls = declarations.filter(ts.isGetAccessor);
-      if (getDecls.length > 0) declarations = getDecls;
-    }
+/** Visitor manages the indexing process for a single TypeScript SourceFile. */
+class Visitor {
+  /** kFile is the VName for the 'file' node representing the source file. */
+  kFile: VName;
 
-    const decl = declarations[0];
-    const vname = this.scopedSignature(decl);
-    // The signature of a value is undecorated.
-    // The signature of a type has the #type suffix.
-    // The signature of a namespace has the #namespace suffix.
-    if (ns === TSNamespace.TYPE) {
-      vname.signature += '#type';
-    } else if (ns === TSNamespace.NAMESPACE) {
-      vname.signature += '#namespace';
-    }
+  /** A shorter name for the rootDir in the CompilerOptions. */
+  sourceRoot: string;
 
-    if (!contextApplies) {
-      // Save it in the appropriate slot in the symbolNames table.
-      if (!vnames) vnames = [null, null, null];
-      vnames[ns] = vname;
-      this.symbolNames.set(sym, vnames);
-    }
+  typeChecker: ts.TypeChecker;
 
-    return vname;
+  constructor(
+      private readonly host: IndexerHost,
+      private file: ts.SourceFile,
+  ) {
+    this.sourceRoot =
+        this.host.program.getCompilerOptions().rootDir || process.cwd();
+
+    this.typeChecker = this.host.program.getTypeChecker();
```

Comment:
I think the visitor should defer entirely to the host now that the host owns a typechecker. `getSymbolAtLocation` can be a method inside the host as well. Two type checkers floating around could get confusing, and the indexer has previously tried to hide the existence of a type checker so that callers of `getSymbolAtLocation` don't need to know it exists at all.

Reply:
The typechecker is owned by the program. The typecheckers in the host and the visitor are just convenience aliases of this. I'm fine with moving `getSymbolAtLocation`, but I don't think its necessary to clean up other uses of the typechecker.

Reply:
Well, if you move `getSymbolAtLocation` into the indexing host you won't need to expose the `typechecker` in the visitor since the host also owns the program.

Reply:
I'm confused as to what you're recommending. Could you clarify?

I've moved `getSymbolAtLocation` into the `IndexerHost`. the `ts.TypeChecker` is still used directly in the `Visitor` for aliasing (`this.typeChecker.getAliasedSymbol`) in two spots. I could move that into the `IndexerHost`, but it seems like overkill.
I also don't understand why it's necessary to cache the typechecker as `this.typeChecker`, on either the `IndexerHost` or the `Visitor`, as the `ts.Program` already caches it. Accessing it as `this[.host].program.getTypeChecker()` will not create a new one if one already exists.

Reply:
That's my bad. I didn't see the other usage of `typechecker` in the Visitor. sgtm :+1: